Tests/MaterialLib/TestMPLCapillaryPressureVanGenuchten.cpp

```diff
@@ -0,0 +1,165 @@
+
+    MPL::VariableArray variable_array;
+    fill(begin(variable_array), end(variable_array),
+         std::numeric_limits<double>::quiet_NaN());
+    ParameterLib::SpatialPosition const pos;
+    double const t = std::numeric_limits<double>::quiet_NaN();
+    double const dt = std::numeric_limits<double>::quiet_NaN();
+
+    double const S_0 = -0.01;
+    double const S_max = 1.01;
+    int const n_steps = 10000;
+    for (int i = 0; i <= n_steps; ++i)
+    {
+        double const S_L = S_0 + i * (S_max - S_0) / n_steps;
+        variable_array[static_cast<int>(MPL::Variable::liquid_saturation)] =
+            S_L;
+
+        double const p_cap =
+            pressure.template value<double>(variable_array, pos, t, dt);
+        ASSERT_LE(p_cap, maximum_capillary_pressure);
+        ASSERT_GE(p_cap, 0);
+
+        variable_array[static_cast<int>(MPL::Variable::capillary_pressure)] =
+            p_cap;
+
+        double const S_L_roundtrip =
+            saturation.template value<double>(variable_array, pos, t, dt);
+
+        if (S_L < residual_liquid_saturation)
+        {
+            EXPECT_GE(S_L_roundtrip, S_L)
+                << "for liquid saturation " << S_L << " and capillary pressure "
+                << p_cap;
+        }
+
+        if (residual_liquid_saturation <= S_L &&
+            S_L <= 1. - residual_gas_saturation &&
+            p_cap < maximum_capillary_pressure)
+        {
+            ASSERT_LE(std::abs(S_L - S_L_roundtrip), 1e-15)
+                << "for liquid saturation " << S_L << " and capillary pressure "
+                << p_cap;
+        }
+
+        if (1. - residual_gas_saturation < S_L)
+        {
+            ASSERT_EQ(S_L_roundtrip, 1. - residual_gas_saturation)
+                << "for liquid saturation " << S_L << " and capillary pressure "
+                << p_cap;
+        }
+
+        double const dp_cap = pressure.template dValue<double>(
+            variable_array, MPL::Variable::liquid_saturation, pos, t, dt);
+        double const dp_cap2 = pressure.template d2Value<double>(
+            variable_array, MPL::Variable::liquid_saturation,
+            MPL::Variable::liquid_saturation, pos, t, dt);
+
+        double const eps = 1e-6;
+        variable_array[static_cast<int>(MPL::Variable::liquid_saturation)] =
+            S_L - eps;
+        double const p_cap_minus =
+            pressure.template value<double>(variable_array, pos, t, dt);
+        variable_array[static_cast<int>(MPL::Variable::liquid_saturation)] =
+            S_L + eps;
+        double const p_cap_plus =
+            pressure.template value<double>(variable_array, pos, t, dt);
+
+        double const Dp_cap = (p_cap_plus - p_cap_minus) / 2 / eps;
+        double const Dp_cap2 =
+            (p_cap_plus - 2 * p_cap + p_cap_minus) / (eps * eps);
+
+        //
+        // First order derivative tests
+        //
+
+        // Testing relative error up to the S_L_max with different tolerances as
```

Comment:
For a smooth continue function, e.g., dp_c(S) of this model, testing it in any segment of the variable range is enough.   Therefore, the test with the smallest tolerance is enough.

Reply:
How do you guarantee, that the function is not called outside of its domain of definition?
How should a Newton method behave if it guesses a value outside of the domain of definition and gets a NaN or -inf as result?

Reply:
The p_c(S) is not smooth, just continuous.

Reply:
I replace `smooth` with ` differentiable` in my comment. With the variable restriction of (S_r, S_m) and value restriction (0, p_c_max), test within (S_r+epsilon, S_m-epsilon) with a relative small tolerance is enough to validate the correction of the implementation. Anyway, the heavy test is also good.

Reply:
The smoothness issue aside, the point of the test is to verify the implementation behaviour outside of the "middle"/well-behaved domain, especially around (in both directions) the residual saturation, the pressure cutoff, and the maximum saturation.